src/rgw/rgw_data_sync.cc

```diff
@@ -1296,7 +1296,8 @@ class RGWDataSyncSingleEntryCR : public RGWCoroutine {
 
         // loop until the latest obligation is satisfied, because other callers
         // may update the obligation while we're syncing
-        while (state->progress_timestamp < state->obligation->timestamp &&
+        while ((state->obligation->timestamp == ceph::real_time() ||
```

Comment:
How does this loop exit if obligation->timestamp is 0?

Reply:
```c++
               obligation_counter != state->counter) {
          obligation_counter = state->counter;
```
this part will break the loop unless another obligation comes in while we're in `RGWRunBucketSourcesSyncCR`